File: Data Structures & Algorithms/count-number-of-islands/submission-4.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        visited = set()
        count = 0
        
        def dfs(x, y):
            # Base case
            if x not in range(len(grid[0])) or y not in range(len(grid)) or (x, y) in visited or grid[y][x] == "0":
                return

            else:
                visited.add((x, y))

                directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
                for dx, dy in directions:
                    dfs(x + dx, y + dy)
        
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if (j, i) not in visited and grid[i][j] == "1":
                    count += 1
                    dfs(j, i)
        
        return count

File: Data Structures & Algorithms/search-for-word/submission-4.js
class Solution {
    /**
     * @param {character[][]} board
     * @param {string} word
     * @return {boolean}
     */
    exist(board, word) {
        const visited = new Set();

        function dfs(row, col, str) {
            if (!str) return true;

            // Get all neighbors that are the next needed letter
            const directions = [[0, 1], [0, -1], [1, 0], [-1, 0]];
            for (const [dx, dy] of directions) {
                const new_row = row + dx;
                const new_col = col + dy;
                const key = `${new_row},${new_col}`;
                if (0 <= new_row && new_row < board.length &&
                    0 <= new_col && new_col < board[0].length &&
                    board[new_row][new_col] === str[0] &&
                    !visited.has(key)) {
                        visited.add(key);
                        if (dfs(new_row, new_col, str.slice(1))) return true;
                        visited.delete(key); // backtrack
                }
            }
            return false;
        }

        // run dfs on every tile
        for (let i = 0; i < board.length; i++) {
            for (let j = 0; j < board[0].length; j++) {
                if (board[i][j] === word[0]) {
                    visited.add(`${i},${j}`);
                    if (dfs(i, j, word.slice(1))) return true;
                    visited.delete(`${i},${j}`);
                }
            }
        }

        // Not found
        return false;
    }
}
